backend-shared/util/log.go

```diff
@@ -1,9 +1,36 @@
 package util
 
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-logr/logr"
+)
+
 const (
 	LogLevel_Debug int = 1
 
 	// LogLevel_Warn is used for unexpected conditions that occur, which might
 	// be an actual error, or might just be the normal functioning of the program.
 	LogLevel_Warn int = -1
 )
+
+type ResourceChangeType string
+
+const (
+	ResourceCreated  ResourceChangeType = "Created"
+	ResourceModified ResourceChangeType = "Modified"
+	ResourceDeleted  ResourceChangeType = "Deleted"
+)
+
+func LogAPIResourceChangeEvent(resourceNamespace string, resourceName string, resource interface{}, resourceChangeType ResourceChangeType, log logr.Logger) {
```

Comment:
Another item I thought of while reviewing the code: for security reasons, we don't want to log the contents of `Secret` resources to the cluster.

So we should be able to add something like this at the top:
```go
	// TODO: We should make sure the resource is not-nil first. See the next review comment on that, below.
	
	secret, isSecret := (resource).(*corev1.Secret)
	if isSecret {
		// Make a copy of the resource, before we modify it
		secret = secret.DeepCopy()

		// Remove the data field, as this contains data that should not be logged.
		secret.Data = map[string][]byte{}

		resource = secret
	}
```

The logic is: if the API is a Secret, remove the `data` field from the Secret (which is where the secret data is stored) before we output it.